P1 wifi gateway 30cbb4c

sakinit opened the pull request:
In commit 89c7648 the P1WifiGateway was reverted back due to not working claims. This request starts with undoing that revert, next re-applying changes that were made since that revert and finally bug-fixing the code.

See also https://github.com/letscontrolit/ESPEasy/commit/0485edc274fc6af80b15cea7b26ce620b40865fd#commitcomment-39220294

sakinit committed:
Revert P1WifiGateway to 30cbb4c to bugfix this more generic code

sakinit committed:
Cherry-pick relevant P1WifiGateway updates since 30cbb4c

Cherry-picked from commits 591988c, 5fe5121, 9c384ee, 1de1350

sakinit committed:
Fix the datagram check to be able to send valid messages again

Tested for DSMR version 4, CRC on, with Domoticz as client

sakinit committed:
Fix the serial reading algorithm

The serial data reading algorithm is 'reverted' to commit 1de1350 to
handle the parsing results within reading loop again.
E.g. the 'start detected' error is logged again.

sakinit committed:
Fix P1WifiGateway webserver start

When the plugin task data is initialized, the execution order is:
 - construct new data (and previously: start webserver)
 - destroy existing data
As the webserver is stopped at destruction of the existing data,
start the webserver after the existing data is destroyed ( and no
longer at construction).

sakinit committed:
Remove temporary comments

TD-er commented:
I made a test build for this PR: [ESPEasy_mega-20200515-38-PR_3091.zip](https://www.dropbox.com/s/i8n15ee0shuocwm/ESPEasy_mega-20200515-38-PR_3091.zip?dl=0)

sakinit commented:
ESP32 needs to be fixed as its WiFiServer does not use REUSEADDR with bind. This results in a server start failure when the server is stopped and started at the same port. Only after some internal time-out it will be able to bind again. I think I add a retry mechanism to fix it, tbc.

TD-er commented:
@sakinit That's very interesting.
Do you have a link to the code where this timeout is used/defined?

sakinit commented:
No I haven't, but AFAIK is it normal tcp/ip behavior. ESP8266 uses REUSEADDR and that should imo be the way to go at ESP32 also.

TD-er commented:
It could explain a number of issues we had when I tried to only have the webserver active when there was a connection.
But I reverted that, as it resulted in a number of issues. One of them being the ESP32 not reacting to HTTP requests after some time.

sakinit commented:
Just checked, adding:
`  int option = 1;`
`  if (setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, &option, sizeof(option)) < 0) {`
`      return;`
`  }`
to the ESP32 WiFiServer (i.e. change in platform code) and the server (re)starts.

sakinit committed:
Use ESPeasySerial

sakinit committed:
Restart P1WifiGateway if webserver start failed

Needed for ESP32 as the WiFiServer needs to time-out before it can
reuse the address as its platform implementation does not use REUSEADDR

sakinit committed:
Optimize serial reading algorithm and move client from plugin into task

sakinit committed:
Fix ESP32 reboot cause at boot/setup

sakinit committed:
Optimize serial in duration

sakinit commented:
The file header of P1WifiGateway states that the gateway code is based on P020 Ser2Net.
Are the "Reset target after boot" and "data forwarding from the WifiClient to the serial-out" left-overs of Ser2Net (as the P1Meter does not receive data)?
If so, both parts imo can be removed and the 'now hidden' status led pin can be shown in the gui and made configurable instead.

sakinit commented:
AFAIK, there are two P1 meters to support (http://domoticx.com/p1-poort-slimme-meter-hardware/), DSMR 2 (9600 7E1) and DSMR 4 (115200 8N1). The GUI can be simplified by limiting the number of configurable options, e.g. making the meter version in the GUI selectable (instead of the serial settings).

TD-er commented on src/_P044_P1WifiGateway.ino:
```diff
@@ -5,127 +5,223 @@
+struct P044_Task : public PluginTaskData_base {
 
-struct P044_data_struct : public PluginTaskData_base {
+  enum ParserState : byte {
```
Maybe even make it into an `enum class ParserState : byte`
This way the compiler always checks if a given parameter is of that type.

sakinit replied:
Ok

TD-er commented on src/_P044_P1WifiGateway.ino:
```diff
@@ -5,127 +5,223 @@
+    gatewayPort = portnumber;
     P1GatewayServer = new WiFiServer(portnumber);
-    if (nullptr != P1GatewayServer) {
+    if (nullptr != P1GatewayServer && WiFi.isConnected()) {
```
This function is only called during init, which is only called when you change settings, or boot.
So if the wifi is not yet connected when init of this task is called, the server will not be started.

You do check it periodically, so is it then still needed during init?
If you remove it from init you will run into the problem I will mention at the `checkServer` function :)

sakinit replied:
I need to know the port number, so call it at init as it used to. Whether the server is started or not depends on the platform used. I explicitly added the check for WiFi connected, not because the server would not start, but because the ESP32 tends to reboot when the WiFi becomes active.
I regard the init state as the state in which to create the objects and get everything ready to roll. As there is no separate start state, also (if possible) get things up and running. From that point of view, the server needs (at least) to be created in the init state.

TD-er replied:
> I explicitly added the check for WiFi connected, not because the server would not start, but because the ESP32 tends to reboot when the WiFi becomes active.

Hmm that's an interesting observation.

TD-er commented on src/_P044_P1WifiGateway.ino:
```diff
@@ -5,127 +5,223 @@
+
+  void checkServer() {
+    if (!serverActive(P1GatewayServer) && WiFi.isConnected()) {
+      P1GatewayServer->close();
```
`serverActive` returns also false if `P1GatewayServer == nullptr`

sakinit replied:
Yes, in that situation the server is also not active.

TD-er replied:
But then the next line will crash the ESP as you try to dereference a null-pointer.

sakinit replied:
Yep, if you consider `P044_Task` as just a bunch of data and methods.
A pointer to an uninitialized task object can only be obtained by `init` and is used only at the `init` state. The `init` state is only successful if the task has been initialized.
In all other cases the task pointer is obtained by method `get` which checks if all pointers are unequal to 0.
Both 'startServer' (of the Gateway) and `serialBegin` (of the Parser) are the means to initialize both parts of the task.
But if you want I can add additional sanity checks...

TD-er replied:
Both are OK.
Either refactoring the code to be part of the struct that on itself is initialized or not, or have a check for null pointer.

The 'problem' with the current code is that it may work correct as it is used now.
But having a separate function that can be called is inviting to be called from other parts of the code.
So a function must not expect external pre-conditions to exist when it should be called or not.
If there are specific conditions on when to call the function, the function itself should check on it, or you must hide it as a private function in an object. As private function of an object, it is up to the object to check.

sakinit replied:
Sanity check added

TD-er replied:
Can you change `WiFi.isConnected()` to `WiFiConnected()` ?

sakinit replied:
Ok, was not aware of that one.

TD-er commented on src/_P044_P1WifiGateway.ino:
```diff
@@ -5,127 +5,223 @@
-    serial_buffer.reserve(P044_BUFFER_SIZE);
-    bytes_read = 0;
+  bool hasClientConnected() {
+    if (P1GatewayServer->hasClient())
```
No check for `nullptr`

sakinit replied:
Already checked for when obtaining the pointer to the task, thus precondition here.

sakinit replied:
Sanity check added

TD-er commented on src/_P044_P1WifiGateway.ino:
```diff
@@ -5,127 +5,223 @@
-  bool isInit() const {
-  	return nullptr != P1GatewayServer;
+  void handleClientIn() {
+    uint8_t net_buf[P044_NETBUF_SIZE];
```
`uint8_t net_buf[P044_NETBUF_SIZE] = {0};`
So it will be initialized with 0 values.
You may still want to 0-terminate it after writing, but at least the received string will also be null-terminated if `net_bytes_read < count`

sakinit replied:
Just moved this code without altering it. See remark in ticket about this part of code being superfluous and imo can be removed (https://github.com/letscontrolit/ESPEasy/pull/3091#issuecomment-635974273).

TD-er replied:
Ah if it can be removed, please do :)

sakinit replied:
Referring explicitly to my question:
> The file header of P1WifiGateway states that the gateway code is based on P020 Ser2Net.
> Are the "Reset target after boot" and "data forwarding from the WifiClient to the serial-out" left-overs of Ser2Net (as the P1Meter does not receive data)?
> If so, both parts imo can be removed and the 'now hidden' status led pin can be shown in the gui and made configurable instead.

Do you know the answer (as my opinion is just what it is: an opinion)?
Especially, the second one is somewhat more tricky as it also impacts the plugin configuration (users needs to reconfigure after upgrade)

TD-er replied:
> Are the "Reset target after boot" and "data forwarding from the WifiClient to the serial-out" left-overs of Ser2Net (as the P1Meter does not receive data)?

I don't know.
The P1 meter is indeed "send only" so there is no need at all for such a communication back.
Only thing you should watch out for is to change the meaning of existing settings out there.
The plugin has been used for quite a while, so there are existing settings out there for this plugin.
In very specific cases we can make a settings transition, based on the "BUILD" value, which is also stored in the settings. But that's only a forward change, so it will then alter the settings and loading an older build again will then use the altered settings.

sakinit replied:
Ok, removed client forwarding data to serial

TD-er commented on src/_P044_P1WifiGateway.ino:
```diff
@@ -5,127 +5,223 @@
+        addLog(LOG_LEVEL_ERROR, F("P1   : Error: network buffer full!"));
+      }
+      net_buf[count] = 0; // before logging as a char array, zero terminate the last position to be safe.
+      char log[P044_NETBUF_SIZE + 40] = {0};
```
`net_buf` and this `log` will both be stack allocated.
Better use `String` for this one (and check if the log level is active), as that one will be heap allocated.
The stack isn't that large and allocating 128 bytes twice + some is quite a lot on the stack.

sakinit replied:
Just moved this code without altering it. See remark in ticket about this part of code being superfluous and imo can be removed.

sakinit replied:
Code removed

TD-er commented on src/_P044_P1WifiGateway.ino:
```diff
@@ -5,127 +5,223 @@
+  }
+
+  void checkBlinkLED() {
+    if (blinkLEDStartTime > 0 && millis() - blinkLEDStartTime >= 500) {
```
Please use `timePassedSince`

sakinit replied:
Ok

sakinit committed:
Update due to review comments

sakinit committed:
Add sanity checks based on review comments

sakinit committed:
Discard data received from WiFi client due to review comment

sakinit commented:
As using `ESPEasySerial `seems the way to go, is the plan to support it in the `PLUGIN_SERIAL_IN` state also?

TD-er commented:
> As using `ESPEasySerial `seems the way to go, is the plan to support it in the `PLUGIN_SERIAL_IN` state also?

Yep and also for the log output.
`ESPEasySerial` also supports an NXP I2C to UART chip, so we can extend the number of UARTs to whatever is needed.
This also means we can redirect all serial related code to use `ESPEasySerial`

I've already changed almost all plugins to use this wrapper and also the RN2483 LoRaWAN controller.

sakinit commented:
Nice

sakinit committed:
Update based on review comments

TD-er committed:
Merge branch 'mega' into P1WifiGateway-30cbb4c

TD-er committed:
Merge branch 'mega' into P1WifiGateway-30cbb4c
